modules/dnn/src/vkcom/src/vkcom.cpp

```diff
@@ -0,0 +1,403 @@
+//
+// Copyright (C) 2018, Intel Corporation, all rights reserved.
+// Third party copyrights are property of their respective owners.
+
+#include "../../precomp.hpp"
+#include "common.hpp"
+#include "internal.hpp"
+#include "../include/op_conv.hpp"
+#include "../include/op_pool.hpp"
+#include "../include/op_lrn.hpp"
+#include "../include/op_concat.hpp"
+#include "../include/op_softmax.hpp"
+#include "../vulkan/vk_loader.hpp"
+
+namespace cv { namespace dnn { namespace vkcom {
+
+#ifdef HAVE_VULKAN
+
+static bool enableValidationLayers = false;
+static VkInstance kInstance;
+static VkPhysicalDevice kPhysicalDevice;
+static VkDebugReportCallbackEXT kDebugReportCallback;
+static uint32_t kQueueFamilyIndex;
+std::vector<const char *> kEnabledLayers;
+typedef std::map<std::thread::id, Context*> IdToContextMap;
+IdToContextMap kThreadResources;
+static std::map<std::string, std::vector<uint32_t>> kShaders;
+static int init_count = 0;
+static bool init();
+static void release();
+static uint32_t getComputeQueueFamilyIndex();
+static bool checkExtensionAvailability(const char *extension_name,
+                                       const std::vector<VkExtensionProperties>
+                                       &available_extensions);
+static VKAPI_ATTR VkBool32 VKAPI_CALL debugReportCallbackFn(
+       VkDebugReportFlagsEXT                       flags,
+       VkDebugReportObjectTypeEXT                  objectType,
+       uint64_t                                    object,
+       size_t                                      location,
+       int32_t                                     messageCode,
+       const char*                                 pLayerPrefix,
+       const char*                                 pMessage,
+       void*                                       pUserData);
+
+static void setContext(Context* ctx)
+{
+    cv::AutoLock lock(getInitializationMutex());
+    std::thread::id tid = std::this_thread::get_id();
+    if (kThreadResources.find(tid) != kThreadResources.end())
+    {
+        return;
+    }
+    kThreadResources.insert(std::pair<std::thread::id, Context*>(tid, ctx));
+}
+
+Context* getContext()
+{
+    Context* ctx = NULL;
+
+    cv::AutoLock lock(getInitializationMutex());
+    std::thread::id tid = std::this_thread::get_id();
+    IdToContextMap::iterator it = kThreadResources.find(tid);
+    if (it != kThreadResources.end())
+    {
+        ctx = it->second;
+    }
+    return ctx;
+}
+
+static void removeContext()
+{
+    cv::AutoLock lock(getInitializationMutex());
+    std::thread::id tid = std::this_thread::get_id();
+    IdToContextMap::iterator it = kThreadResources.find(tid);
+    if (it == kThreadResources.end())
+    {
+        return;
+    }
+    kThreadResources.erase(it);
+}
+
+bool initPerThread()
+{
+    VkDevice device;
+    VkQueue queue;
+    VkCommandPool cmd_pool;
+
+    VKCOM_CHECK_BOOL_RET_VAL(init(), false);
+    Context* ctx = getContext();
+    if (ctx)
+    {
+        ctx->ref++;
```

Comment:
`ctx` is "thread local"?
If it is not, then the reference counted should be synchronized between threads properly (via `CV_XADD()` macro or use `std::atomic<int>` (OpenCV 4.0 requires C++11) ).

Reply:
Yes, it is thread local. 
It is for multiple callings on initPerThread(). e.g, multiple Net objects exsiting on a thread simultaneously. 

I am looking into the issues you and @dkurt mentioned. But I have short time on coding in these days and next week for I will be in biz travel. Responses and fixes may be a little bit late. Sorry.